File: typescript/src/BestValueMenu.ts
import {Notifier} from "./Notifier";

export class BestValueMenu {
    private notifier;
    constructor(notifier: Notifier) {
        this.notifier = notifier;
    }


    serveMenu(): string {
        return this.starter() + "\n"
        + this.mainCourse() + "\n"
        + this.dessert()
    }

    private starter() {
        return "sallad";
    }

    protected mainCourse() {
        this.allergenNotificiation("quiche lorraine", "milk")
        return "quiche lorraine";
    }

    protected dessert() {
        this.allergenNotificiation("brownie", "nuts")
        return "brownie";
    }

    protected allergenNotificiation(plate: string, allergen: string) {
        this.notifier.log(`[Allergen]: ${plate} contains ${allergen}` )
    }
}

File: java/src/main/java/inheritance/BestValueMenu.java
package inheritance;

public class BestValueMenu {

    private final Dessert dessert;
    private final MainCourse mainCourse;
    private final AllergenNotifier allergenNotifier;

    public BestValueMenu(Dessert dessert, MainCourse mainCourse, AllergenNotifier allergenNotifier) {
        this.dessert = dessert;
        this.mainCourse = mainCourse;
        this.allergenNotifier = allergenNotifier;
    }

    public String serveMenu() {
        allergenNotifier.allergenNotification(mainCourse.mainCourse(), concatAllergens(mainCourse));
        allergenNotifier.allergenNotification(dessert.dessert(), concatAllergens(dessert));
        return starter() + "\n" +
                mainCourse.mainCourse() + "\n" +
                dessert.dessert();
    }

    private String concatAllergens(Dish dish) {
        return String.join(", ", dish.allergens());
    }

    private String starter() {
        return "sallad";
    }

}
